Source/NSOrderedSet.m

```diff
@@ -0,0 +1,1814 @@
+}
+
+- (void) encodeWithCoder: (NSCoder *)aCoder
+{
+   if ([aCoder allowsKeyedCoding])
+    {
+      NSMutableArray *array = [NSMutableArray array];
+      NSEnumerator *en = [self objectEnumerator];
+      id obj = nil;
+      /* HACK ... MacOS-X seems to code differently if the coder is an
+       * actual instance of NSKeyedArchiver
+       */
+      
+      // Collect all objects...
+      while((obj = [en nextObject]) != nil)
+	{
+	  [array addObject: obj];
+	}
+      
+      if ([aCoder class] == [NSKeyedArchiver class])
+	{
+	  [(NSKeyedArchiver*)aCoder _encodeArrayOfObjects: array
+						   forKey: @"NS.objects"];
+	}
+      else
+	{
+	  unsigned	i = 0;
+	  NSEnumerator	*e = [self objectEnumerator];
+	  id		o;
+
+	  while ((o = [e nextObject]) != nil)
+	    {
+	      NSString	*key;
+
+	      key = [NSString stringWithFormat: @"NS.object.%u", i++];
+	      [(NSKeyedArchiver*)aCoder encodeObject: o forKey: key];
+	    }
+	}
+    }
+  else
+    {
+      unsigned		count = [self count];
+      NSEnumerator	*e = [self objectEnumerator];
+      id		o;
+
+      [aCoder encodeValueOfObjCType: @encode(unsigned) at: &count];
+      while ((o = [e nextObject]) != nil)
+	{
+	  [aCoder encodeValueOfObjCType: @encode(id) at: &o];
+	}
+    } 
+}
+
+- (id) copyWithZone: (NSZone*)zone
+{
+  NSOrderedSet	*copy = [NSOrderedSet_concrete_class allocWithZone: zone];
+  return [copy initWithOrderedSet: self copyItems: YES];
+}
+
+- (id) mutableCopyWithZone: (NSZone*)zone
+{
+   NSMutableOrderedSet	*copy = [NSMutableOrderedSet_concrete_class allocWithZone: zone];
+   return [copy initWithOrderedSet: self copyItems: NO];
+}
+
+// NSFastEnumeration 
+- (NSUInteger) countByEnumeratingWithState: (NSFastEnumerationState *)state
+				   objects: (__unsafe_unretained id[])stackbuf
+				     count: (NSUInteger)len
+{
+  NSInteger count;
+
+  /* In a mutable subclass, the mutationsPtr should be set to point to a
+   * value (unsigned long) which will be changed (incremented) whenever
+   * the container is mutated (content added, removed, re-ordered).
+   * This is cached in the caller at the start and compared at each
+   * iteration.   If it changes during the iteration then
+   * objc_enumerationMutation() will be called, throwing an exception.
+   * The abstract base class implementation points to a fixed value
+   * (the enumeration state pointer should exist and be unchanged for as
+   * long as the enumeration process runs), which is fine for enumerating
+   * an immutable array.
+   */
+  state->mutationsPtr = (unsigned long *)&state->mutationsPtr;
+  count = MIN(len, [self count] - state->state);
+  /* If a mutation has occurred then it's possible that we are being asked to
+   * get objects from after the end of the array.  Don't pass negative values
+   * to memcpy.
+   */
+  if (count > 0)
+    {
+      IMP	imp = [self methodForSelector: @selector(objectAtIndex:)];
+      int	p = state->state;
+      int	i;
+
+      for (i = 0; i < count; i++, p++)
```

Comment:
You could actually use [self getObjects: stackbuf range: NSMakeRange(state->state, count)]; here and benefit from a better implementation of that method in a subclass.